modules/oas-page-builder/src/services/pageBuilder.ts

```diff
@@ -50,39 +65,76 @@ const getAtlasSpecUrl = async (apiKeyword: string) => {
   return oasFileURL;
 };
 
+interface GetOASpecParams {
+  sourceType: string;
+  source: string;
+  output: string;
+  pageSlug: string;
+  repoPath: string;
+  redocExecutor: RedocExecutor;
+  apiVersion?: string;
+  resourceVersion?: string;
+}
+
+async function getOASpec({
+  source,
+  sourceType,
+  repoPath,
+  pageSlug,
+  redocExecutor,
+  output,
+  apiVersion,
+  resourceVersion,
+}: GetOASpecParams) {
+  try {
+    let spec = '';
+    const buildOptions: RedocBuildOptions = {};
+
+    if (sourceType === 'url') {
+      spec = source;
+    } else if (sourceType === 'local') {
+      const localFilePath = normalizePath(`${repoPath}/source/${source}`);
+      spec = localFilePath;
+    } else if (sourceType === 'atlas') {
+      spec = await getAtlasSpecUrl({ apiKeyword: source, apiVersion, resourceVersion });
+      // Ignore "incompatible types" warnings for Atlas Admin API/cloud-docs
+      buildOptions['ignoreIncompatibleTypes'] = true;
+    } else {
+      throw new Error(`Unsupported source type "${sourceType}" for ${pageSlug}`);
+    }
+
+    const finalFilename = normalizePath(`${output}/${pageSlug}/index.html`);
```

Comment:
For versioned OpenAPI pages, we should have the filename output path include the resource version. For example, if "2023-01-01" is a resource version for v2, the final filename should be something like:

`/reference/api-resources-spec/v2/2023-01-01/index.html`